File: ASSIGNMENT_ARules_movies.py
# ASSIGNMENT ASSOCIATION_RULES my_movies dataset

# importing important libraries
import pandas as pd
import numpy as np
pd.set_option('display.max_columns',500)
import matplotlib.pyplot as plt
from mlxtend.frequent_patterns import apriori, association_rules

movies = pd.read_csv("E:/Association_Rules/my_movies.csv")
movies = movies.iloc[:,5:]
movies

#Quick look at the top items
movies.sum().to_frame('Frequency').sort_values('Frequency',ascending=False).plot(kind='bar',figsize=(12,8),title='Frequent Items')

# getting frequent items
frequent_items = apriori(movies,  min_support=0.2,use_colnames=True)
frequent_items.shape # 13
frequent_items.sort_values('support',ascending=False) 

# Getting rules from the frequent itemsets
rules=association_rules(frequent_items, metric='lift',min_threshold=1)
rules.shape # we have obtained 16 rules 
rules.head()

# removing redundant rules
def sort_rules(i):
    return sorted(list(i))
rules_movies = rules.antecedents.apply(sort_rules) + rules.consequents.apply(sort_rules)
rules_movies = rules_movies.apply(sorted)   
rules_sets = list(rules_movies)
unique_rules_sets = [list(m) for m in set(tuple(i) for i in rules_sets)]
index_rules = []
for i in unique_rules_sets:
    index_rules.append(rules_sets.index(i))

# getting rules without any redundancy 
rules_no_redundancy  = rules.iloc[index_rules,:]
rules.shape # we have 16 rules including duplicate rules
rules_no_redundancy.shape # now we have 6 rules after removing duplicate rules

# Sorting them with respect to lift 
rules_no_redundancy = rules_no_redundancy.sort_values('lift',ascending=False)

#### By changing the metric to confidence
rules1 = association_rules(frequent_items,metric='confidence',min_threshold=0.8)
# by changing metric to 0.8 confidence we get 8 rules and if remove duplicate rules
# we get 5 rules

#############################
%matplotlib qt
def draw_graph(rules_no_redundancy, rules_to_show):
  import networkx as nx  
  G1 = nx.DiGraph()
   
  color_map=[]
  N = 50
  colors = np.random.rand(N)    
  strs=['R0', 'R1', 'R2', 'R3', 'R4', 'R5', 'R6', 'R7', 'R8', 'R9', 'R10', 'R11']   
   
   
  for i in range (rules_to_show):      
    G1.add_nodes_from(["R"+str(i)])
    
     
    for a in rules_no_redundancy.iloc[i]['antecedents']:
                
        G1.add_nodes_from([a])
        
        G1.add_edge(a, "R"+str(i), color=colors[i] , weight = 2)
       
    for c in rules_no_redundancy.iloc[i]['consequents']:
             
            G1.add_nodes_from([c])
            
            G1.add_edge("R"+str(i), c, color=colors[i],  weight=2)
 
  for node in G1:
       found_a_string = False
       for item in strs: 
           if node==item:
                found_a_string = True
       if found_a_string:
            color_map.append('yellow')
       else:
            color_map.append('green')       
 
 
   
  edges = G1.edges()
  colors = [G1[u][v]['color'] for u,v in edges]
  weights = [G1[u][v]['weight'] for u,v in edges]
 
  pos = nx.spring_layout(G1, k=16, scale=1)
  nx.draw(G1, pos, edges=edges, node_color = color_map, edge_color=colors, width=weights, font_size=16, with_labels=False)            
   
  for p in pos:  # raise text positions
           pos[p][1] += 0.07
  nx.draw_networkx_labels(G1, pos)
  plt.show()
 
     
draw_graph (rules_no_redundancy, 1) 
draw_graph(rules_no_redundancy,6)
rules_no_redundancy

draw_graph(rules_no_redundancy,2)
'''
yellow circle gives the rule name.. R0 is the first rule
If arrow is away from the green circles , then that is antecedent, here LOTR1
if arrow is towards the green circle, then that is consequent, here LOTR2
'''

####### RECOMMENDATIONS
'''
1. The first rule is LOTR1 ==> LOTR2
    if anyone has watched either of the movie (not both), then we can 
    recommend them to watch the other sequel
2. Those who have watched either Green-Mile or Sixth-Sense, we can recommend
    them to watch the other movie
3. If watched one or two of the moives Sixth-Sense, Gladiator and Patriot,
    we can recommend them to watch the other movies
'''

File: ASSIGNMENT_ARules_movies.R
##################### ASSIGNMENT ASSOCIATION RULES on MOVIES dataset

library(arules)
library(arulesViz)

movies_raw <- read.csv('E:\\Association_Rules\\my_movies.csv')
head(movies_raw)
movies <- as.matrix(movies_raw[,-c(1,2,3,4,5)])
head(movies)
class(movies)

rules_movie <- apriori(movies,parameter=list(support=0.2,confidence=0.4))
# sup 0.22, conf 0.4 get 15 rules
# when do not give min length then get some rules with only consequent (no antecedent)
inspect(rules_movie)

rules_movie <- apriori(movies,parameter=list(support=0.1,confidence=0.4,minlen=2))
# sup 0.1, conf 0.4, minlen 2 , get 105 rules
# sup 0.1 indicates that movie should be present in 10% of transactions. since there
# are only 10 transactions, this implies that even if movie is present in only
# 1 transaction, the movie will be used to generate rules. This is not proper
inspect(rules_movie)

# remove redundant rules
'''
the first rule is  Harry.Potter2 => Harry.Potter1
the second rule is Harry.Potter1 => Harry.Potter2
both are same rules, we need to remove one duplicate rule
'''
rules_movie[is.redundant(rules_movie)] # there are 58 redundant rules
rules_movie_redund <- rules_movie[!is.redundant(rules_movie)] # removing redundant rules
rules_movie_redund # get 47 rules after removing 58 duplicate rules

rules_movie1 <- apriori(movies,parameter=list(support=0.2,confidence=0.4))
# sup 0.2, conf 0.4 get 15 rules
# when do not give min length then get some rules with only consequent (no antecedent)
inspect(rules_movie1)

rules_movie2 <- apriori(movies,parameter=list(support=0.2,confidence=0.4,minlen=2))
# sup 0.2, conf 0.4 get 108 rules
inspect(rules_movie2)

rules_movie2_df = data.frame(
  lhs = labels(lhs(rules_movie2)),
  rhs = labels(rhs(rules_movie2)), 
  rules_movie2@quality)

library(writexl)
write_xlsx(rules_movie2_df,'E:\\Association_Rules\\rules_movies.xlsx')
